File: api/postauth.py
"""This module provides the post-auth logic"""

import logging
from datetime import datetime
from urllib import parse
from .constants import POSTAUTH_LINE, POSTAUTH_LOG_FILE
from .models import Machine, User, BaseResult
from .exceptions import MachineNotFoundException, NoMoreIPException, ReadOnlyException
from .messages import PostAuthMessage as Message


class Result(BaseResult):
    """This class represents a post-auth result"""

    def is_ok(self):
        """
        Return whether the full authentication is complete or not.
        :returns: A boolean telling if the client is authenticated
        """
        return self.message.value == 0


def with_mac(ldap, mac, user_name):
    """
    Do a MAC post-auth.
    :param ldap: The ldap to connect to
    :param mac: The client's MAC address
    :param user_name: The user name passed by the NAS (should be the same as the MAC address)
    """
    if user_name != mac:
        logging.warning('[POSTAUTH][with_mac] ({}) failed: inconsistent MAC'.format(mac))
        return Result('MAC', Message.INCONSISTENT_MAC)

    try:
        machine_data = ldap.get_machine(mac)
    except MachineNotFoundException:
        logging.error('[POSTAUTH][with_mac] ({}) failed: unregistered device'.format(mac))
        return Result('MAC', Message.UNREGISTERED_MACHINE)
    machine = Machine(ldap, **machine_data)
    try:
        machine.update_last_date()
    except:
        pass
    return Result('MAC', Message.OK, machine)


def with_dot1x(ldap, mac, user_name):
    """
    Do a post-auth using 802.1X data.
    :param ldap: The ldap to connect to
    :param mac: The client's MAC address
    :param user_name: The user name passed by the NAS
    """
    try:
        user = User(**ldap.get_user(user_name))
        machine_data = None
        try:
            machine_data = ldap.get_machine(mac)
        except MachineNotFoundException:
            try:
                ldap.add_machine(mac, user)
                machine_data = ldap.get_machine(mac)
            except ReadOnlyException:
                logging.error('[POSTAUTH][with_dot1x] ({}*{}) failed: readonly LDAP'
                              .format(user_name, mac))
                return Result('802.1X', Message.LDAP_ERROR)
        machine = Machine(ldap, **machine_data)
        try:
            machine.update_last_date()
        except:
            pass
        return Result('802.1X', Message.OK, machine)
    except NoMoreIPException:
        logging.critical('[POSTAUTH][with_dot1x] ({}*{}) failed: all nodes DOWN'
                         .format(user_name, mac))
        return Result('802.1X', Message.LDAP_ERROR)


def wrongpassword(ldap, mac, user_name):
    """
    Do a post-auth with a wrong password.
    :param ldap: The ldap to connect to
    :param mac: The client's MAC address
    :param user_name: The user name passed by the NAS
    """
    machine = None
    try:
        machine_data = ldap.get_machine(mac)
        machine = Machine(ldap, **machine_data)
    except MachineNotFoundException:
        pass
    except NoMoreIPException:
        logging.critical('[WRONGPASSWORD][process] ({}*{}) failed: all nodes DOWN'
                         .format(user_name, mac))
        return Result('802.1X', Message.LDAP_ERROR)
    return Result('802.1X', Message.WRONG_PASSWORD, machine)


def log(ip, port, mac, uid, result):
    """
    Save the post-auth result.
    :param ip: The NAS IP
    :param port: The NAS physical port
    :param mac: The client MAC
    :param uid: The client UID
    :param result: The post-auth result
    """
    with open(POSTAUTH_LOG_FILE, 'a') as logfile:
        logfile.write(POSTAUTH_LINE.format(int(datetime.now().timestamp() * 1000000), ip,
                                           parse.quote(port, safe=''), mac,
                                           parse.quote(uid, safe=''),
                                           result.get_machine_owner(), result.message.name,
                                           result.auth))

File: lib/post_auth.ml
(** This module provides the post-authentication logic *)


open Autx
module Log = Dolog.Log


(** Post-authentication functor *)
module With (M : S) = struct
  (** Save the post-authentication result.
      @param ip: The NAS IP
      @param port: The NAS logical port
      @param uid: The client UID
      @param mac: The client MAC
      @param owner: The device owner
      @param message: The authorization message *)
  let log ip port uid mac owner message =
    Util.append_file Constants.postauth_log_file
                     [%string "%{Util.now_us () # Int}// radius.postauth{ip=%{ip},\
                               port=%{Util.percent port},mac=%{mac},uid=%{uid},owner=%{owner},\
                               status=%{Message.to_string message},\
                               auth=%{to_string M.ty}} 1\n"]

  (** Process a post-authentication request
      @param ldap The LDAP connection
      @param ip The NAS IP
      @param port The NAS logical port
      @param uid The client UID
      @param mac The client MAC *)
  let process ldap ip port uid mac =
    let (message, owner) = M.process ldap uid mac in
    log ip port uid mac owner message; message
end


(** MAC post-authentication module *)
module Mac : S = struct
  let ty = Mac

  (** Perform a MAC post-authentication.
      @param ldap The LDAP connection
      @param mac The client's MAC address *)
  let process ldap _ mac =
    let prefix = "[POSTAUTH][with_mac]" in
    match Device.get ldap mac with
      | Ok device ->
          Device.update_date ldap mac;
          Log.info "%s" [%string "%{prefix} (%{mac}) done"];
          (Message.Auth_ok, device.owner_uid)
      | Error Device.Not_found ->
          Log.warn "%s" [%string "%{prefix} (%{mac}) failed: unregistered device"];
          (Message.Unregistered_machine, Owner.unknown)
      | _ -> failwith "Internal error"
end

(** 802.1X post-authentication module *)
module Dot1x : S = struct
  let ty = Dot1x

  (** Perform a post-authentication using 802.1X data.
      @param ldap: The LDAP connection
      @param uid: The user name
      @param mac: The client's MAC address *)
  let process ldap uid mac =
    let prefix = "[POSTAUTH][with_dot1x]" in
    let device = match Device.get ldap mac with
      | Ok device -> device
      | Error (Device.Not_found) ->
          Device.create ldap mac uid;
          Result.get_ok (Device.get ldap mac)
      | _ -> failwith "Internal error" in
    Device.update_date ldap mac;
    Log.info "%s" [%string "%{prefix} (%{mac}) done"];
    (Message.Auth_ok, device.owner_uid)
end

(** Module to handle wrong passwords *)
module Wrong_password : S = struct
  let ty = Dot1x

  (** Handle a wrong password.
      @param ldap: The LDAP connection
      @param mac: The client's MAC address *)
  let process ldap _ mac =
    match Device.get ldap mac with
    | Ok device -> (Message.Wrong_password, device.owner_uid)
    | Error (Device.Not_found) -> (Message.Unregistered_machine, Owner.unknown)
    | _ -> failwith "Internal error"
end
